pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/groupby/DefaultGroupKeyGenerator.java

```diff
@@ -365,20 +562,89 @@ public boolean hasNext() {
 
-      String stringGroupKey = rawKeyToStringGroupKey(groupKey);
-      _groupKey.setFirst(groupId);
+      String stringGroupKey = rawKeyToStringGroupKey(entry.getKey());
+      _groupKey.setFirst(entry.getIntValue());
       _groupKey.setSecond(stringGroupKey);
       return _groupKey;
     }
 
     @Override
     public void remove() {
-      throw new RuntimeException("Unsupported remove method.");
+      throw new UnsupportedOperationException();
     }
   }
+
+  /**
+   * With an integer group key, convert group key from dictId based to string based, using actually values corresponding
+   * to dictionary id's.
+   * (ARRAY_BASED storage type)
+   *
+   * @param groupKey integer group key.
+   * @return string group key.
+   */
+  private String groupKeyToStringGroupKey(int groupKey) {
+    if (_numGroupByColumns == 1) {
+      // Special case one group-by column for performance.
+      return _dictionaries[0].get(groupKey).toString();
+    } else {
+      // Decode the group key.
+      int cardinality = _cardinalities[0];
+      StringBuilder builder = new StringBuilder(_dictionaries[0].get(groupKey % cardinality).toString());
+      groupKey /= cardinality;
+      for (int i = 1; i < _numGroupByColumns; i++) {
+        builder.append(GroupByConstants.GroupByDelimiter.groupByMultiDelimeter);
+        cardinality = _cardinalities[i];
+        builder.append(_dictionaries[i].get(groupKey % cardinality));
+        groupKey /= cardinality;
+      }
+      return builder.toString();
+    }
+  }
+
+  /**
+   * With a long raw key, convert raw key from dictId based to string based group key, using actually values
+   * corresponding to dictionary id's.
+   * (LONG_BASED storage type)
+   *
+   * @param rawKey long raw key.
+   * @return string group key.
+   */
+  private String rawKeyToStringGroupKey(long rawKey) {
+    if (_numGroupByColumns == 1) {
+      // Special case one group-by column for performance.
+      return _dictionaries[0].get((int) rawKey).toString();
+    } else {
+      // Decode the raw key.
+      int cardinality = _cardinalities[0];
+      StringBuilder builder = new StringBuilder(_dictionaries[0].get((int) (rawKey % cardinality)).toString());
+      rawKey /= cardinality;
+      for (int i = 1; i < _numGroupByColumns; i++) {
+        builder.append(GroupByConstants.GroupByDelimiter.groupByMultiDelimeter);
+        cardinality = _cardinalities[i];
+        builder.append(_dictionaries[i].get((int) (rawKey % cardinality)));
+        rawKey /= cardinality;
+      }
+      return builder.toString();
+    }
+  }
+
+  /**
+   * With a IntArrayList raw key, convert raw key from dictId based to string based group key, using actually values
+   * corresponding to dictionary id's.
+   * (ARRAY_MAP_BASED storage type)
+   *
+   * @param rawKey IntArrayList raw key.
+   * @return string group key.
+   */
+  private String rawKeyToStringGroupKey(IntArrayList rawKey) {
```

Comment:
Isn't it butter to let the compiler decide the right function depending on the argument type? (like most lib class does)